PetscScalar conversion errors when Complex type enabled

As of 8e353b8, I am experiencing build errors in `petsc_vector.h` and `petsc_dm_wrapper.C` when `--enable-complex` is used as a configure flag. Just in case this is a bug (hopefully just a configuration error on my part...) I wanted to submit a bug report. 

My machine configuration info is below. **Also of note** - I built PETSc and libMesh as MOOSE submodules using the `update_and_rebuild_` MOOSE scripts with modified flags (I am exploring what it would take to properly configure PETSc and libMesh for a possible future complex-compatible version of MOOSE....the MOOSE Team had a recent user question about it and I'd like it for my own electromagnetics research using MOOSE).

OS: Mojave 10.14.3
PETSc version: 3.10.2
Compiler: clang 6.0.1

libMesh configuration flags:
`--with-methods=opt oprof dbg --prefix=/Users/icenct/projects/moose/scripts/../libmesh/installed --enable-silent-rules --enable-unique-id --disable-warnings --with-thread-model=openmp --disable-maintainer-mode --enable-metaphysicl-required --enable-complex`

PETSc configuration flags:
`--download-hypre=0 --with-ssl=0 --with-debugging=no  --with-pic=1 --with-shared-libraries=1 --with-cc=mpicc --with-cxx=mpicxx --with-fc=mpif90 --download-fblaslapack=1 --download-metis=1 --download-ptscotch=1 --download-parmetis=1 --download-superlu_dist=1 --download-mumps=1 --download-scalapack=1 --CC=mpicc --CXX=mpicxx --FC=mpif90 --F77=mpif77 --F90=mpif90 --CFLAGS='-fPIC -fopenmp' --CXXFLAGS='-fPIC -fopenmp' --FFLAGS='-fPIC -fopenmp' --FCFLAGS='-fPIC -fopenmp' --F90FLAGS='-fPIC -fopenmp' --F77FLAGS='-fPIC -fopenmp'`

They were mostly Number to Real/double conversion errors. I wasn't sure how to include all of the error messages in my original posting, but I figured it out below. I built this on my laptop at home today and reproduced the same issues (on a Mojave MacBook Pro, same configuration as the desktop). One additional error in `mesh_function.C` also appeared.

```
../../tests/mesh/mesh_function.C:123:14: error: no viable conversion from 'libMesh::Number' (aka 'complex<double>') to 'libMesh::Real' (aka 'double')
        Real mesh_function_value = projection_function(*node,
             ^                     ~~~~~~~~~~~~~~~~~~~~~~~~~~

../src/solvers/petsc_dm_wrapper.C:611:36: error: non-const lvalue reference to type 'SparseMatrix<libMesh::Number>' (aka 'SparseMatrix<complex<double> >') cannot bind to a value of unrelated type 'PetscMatrix<libMesh::Real>' (aka 'PetscMatrix<double>')
          system.projection_matrix(*_ctx_vec[i-1]->K_interp_ptr);
                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
./include/libmesh/system.h:1675:50: note: passing argument to parameter 'proj_mat' here
  void projection_matrix (SparseMatrix<Number> & proj_mat) const;
                                                 ^
In file included from ../src/solvers/petsc_dm_wrapper.C:31:
In file included from ./include/libmesh/petsc_dm_wrapper.h:39:
./include/libmesh/petsc_vector.h:1080:10: error: cannot convert 'const PetscScalar' (aka 'const complex<double>') to 'double' without a conversion operator
  return static_cast<T>(_read_only_values[local_index]);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./include/libmesh/petsc_vector.h:72:3: note: in instantiation of member function 'libMesh::PetscVector<double>::operator()' requested here
  PetscVector (const Parallel::Communicator & comm_in,
  ^
/opt/moose/llvm-6.0.1/include/c++/v1/memory:3078:32: note: in instantiation of member function 'libMesh::PetscVector<double>::PetscVector' requested here
    return unique_ptr<_Tp>(new _Tp(std::__1::forward<_Args>(__args)...));
                               ^
../src/solvers/petsc_dm_wrapper.C:1027:26: note: in instantiation of function template specialization 'std::__1::make_unique<libMesh::PetscVector<double>, const libMesh::Parallel::Communicator &>' requested here
      _vec_vec[i] = std::make_unique<PetscVector<Real>>(comm);
                         ^
In file included from ../src/solvers/petsc_dm_wrapper.C:31:
In file included from ./include/libmesh/petsc_dm_wrapper.h:39:
./include/libmesh/petsc_vector.h:1103:19: error: cannot convert 'const PetscScalar' (aka 'const complex<double>') to 'double' without a conversion operator
      values[i] = static_cast<T>(_read_only_values[local_index]);
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./include/libmesh/petsc_vector.h:72:3: note: in instantiation of member function 'libMesh::PetscVector<double>::get' requested here
  PetscVector (const Parallel::Communicator & comm_in,
  ^
/opt/moose/llvm-6.0.1/include/c++/v1/memory:3078:32: note: in instantiation of member function 'libMesh::PetscVector<double>::PetscVector' requested here
    return unique_ptr<_Tp>(new _Tp(std::__1::forward<_Args>(__args)...));
                               ^
../src/solvers/petsc_dm_wrapper.C:1027:26: note: in instantiation of function template specialization 'std::__1::make_unique<libMesh::PetscVector<double>, const libMesh::Parallel::Communicator &>' requested here
      _vec_vec[i] = std::make_unique<PetscVector<Real>>(comm);
                         ^
```

--- a/examples/fem_system/fem_system_ex1/fem_system_ex1.C
+++ b/examples/fem_system/fem_system_ex1/fem_system_ex1.C
@@ -72,6 +72,12 @@ int main (int argc, char ** argv)
   libmesh_example_requires(false, "--disable-singleprecision");
 #endif
 
+  // This example uses the new PetscDMWrapper which currently does not
+  // compile when complex numbers are enabled.
+#ifdef LIBMESH_USE_COMPLEX_NUMBERS
+  libmesh_example_requires(false, "--disable-complex");
+#endif
+
 #ifndef LIBMESH_ENABLE_AMR
   libmesh_example_requires(false, "--enable-amr");
 #else

--- a/include/solvers/petsc_diff_solver.h
+++ b/include/solvers/petsc_diff_solver.h
@@ -97,10 +97,8 @@ protected:
    * Wrapper object for interacting with PetscDM
    */
 #if !PETSC_VERSION_LESS_THAN(3,7,3)
-#ifdef LIBMESH_ENABLE_AMR
-#ifdef LIBMESH_HAVE_METAPHYSICL
+#if defined(LIBMESH_ENABLE_AMR) && defined(LIBMESH_HAVE_METAPHYSICL) && !defined(LIBMESH_USE_COMPLEX_NUMBERS)
   PetscDMWrapper _dm_wrapper;
-#endif
 #endif
 #endif
 

--- a/include/solvers/petsc_dm_wrapper.h
+++ b/include/solvers/petsc_dm_wrapper.h
@@ -23,8 +23,7 @@
 
 #ifdef LIBMESH_HAVE_PETSC
 #if !PETSC_VERSION_LESS_THAN(3,7,3)
-#ifdef LIBMESH_ENABLE_AMR
-#ifdef LIBMESH_HAVE_METAPHYSICL
+#if defined(LIBMESH_ENABLE_AMR) && defined(LIBMESH_HAVE_METAPHYSICL) && !defined(LIBMESH_USE_COMPLEX_NUMBERS)
 
 #include <vector>
 #include <memory>
@@ -219,8 +218,7 @@ private:
 
 }
 
-#endif // #if LIBMESH_HAVE_METAPHYSICL
-#endif // #if LIBMESH_ENABLE_AMR
+#endif // #if LIBMESH_ENABLE_AMR && LIBMESH_HAVE_METAPHYSICL && !LIBMESH_USE_COMPLEX_NUMBERS
 #endif // #if PETSC_VERSION
 #endif // #ifdef LIBMESH_HAVE_PETSC
 

--- a/src/solvers/petsc_diff_solver.C
+++ b/src/solvers/petsc_diff_solver.C
@@ -228,10 +228,8 @@ void PetscDiffSolver::clear()
   LIBMESH_CHKERR(ierr);
 
 #if !PETSC_VERSION_LESS_THAN(3,7,3)
-#ifdef LIBMESH_ENABLE_AMR
-#ifdef LIBMESH_HAVE_METAPHYSICL
+#if defined(LIBMESH_ENABLE_AMR) && defined(LIBMESH_HAVE_METAPHYSICL) && !defined(LIBMESH_USE_COMPLEX_NUMBERS)
   _dm_wrapper.clear();
-#endif
 #endif
 #endif
 }
@@ -370,11 +368,9 @@ void PetscDiffSolver::setup_petsc_data()
 
   // This needs to be called before SNESSetFromOptions
 #if !PETSC_VERSION_LESS_THAN(3,7,3)
-#ifdef LIBMESH_ENABLE_AMR
-#ifdef LIBMESH_HAVE_METAPHYSICL
+#if defined(LIBMESH_ENABLE_AMR) && defined(LIBMESH_HAVE_METAPHYSICL) && !defined(LIBMESH_USE_COMPLEX_NUMBERS)
   if (use_petsc_dm)
     this->_dm_wrapper.init_and_attach_petscdm(_system, _snes);
-#endif
 #endif
 #endif
 

--- a/src/solvers/petsc_dm_wrapper.C
+++ b/src/solvers/petsc_dm_wrapper.C
@@ -20,8 +20,7 @@
 
 #ifdef LIBMESH_HAVE_PETSC
 #if !PETSC_VERSION_LESS_THAN(3,7,3)
-#ifdef LIBMESH_ENABLE_AMR
-#ifdef LIBMESH_HAVE_METAPHYSICL
+#if defined(LIBMESH_ENABLE_AMR) && defined(LIBMESH_HAVE_METAPHYSICL) && !defined(LIBMESH_USE_COMPLEX_NUMBERS)
 
 #include "libmesh/ignore_warnings.h"
 #include <petscsf.h>
@@ -1030,7 +1029,6 @@ void PetscDMWrapper::init_dm_data(unsigned int n_levels, const Parallel::Communi
 
 } // end namespace libMesh
 
-#endif // LIBMESH_HAVE_METAPHYSICL
-#endif // LIBMESH_ENABLE_AMR
+#endif // #if LIBMESH_ENABLE_AMR && LIBMESH_HAVE_METAPHYSICL && !LIBMESH_USE_COMPLEX_NUMBERS
 #endif // PETSC_VERSION
 #endif // LIBMESH_HAVE_PETSC

--- a/tests/mesh/mesh_function.C
+++ b/tests/mesh/mesh_function.C
@@ -120,12 +120,16 @@ public:
     for (const auto & node : mesh.local_node_ptr_range())
       {
         (*mesh_function)(*node, /*time=*/ 0., vec_values);
-        Real mesh_function_value = projection_function(*node,
-                                                       es.parameters,
-                                                       dummy,
-                                                       dummy);
+        Number mesh_function_value =
+          projection_function(*node,
+                              es.parameters,
+                              dummy,
+                              dummy);
 
-        CPPUNIT_ASSERT_DOUBLES_EQUAL(vec_values(0), mesh_function_value, TOLERANCE*TOLERANCE);
+        CPPUNIT_ASSERT_DOUBLES_EQUAL
+          (libmesh_real(vec_values(0)),
+           libmesh_real(mesh_function_value),
+           TOLERANCE*TOLERANCE);
       }
   }
 #endif // LIBMESH_ENABLE_AMR

